drivers/timer/rk3568/timer.c

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright 2025, UNSW
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <stdint.h>
+#include <os/sddf.h>
+#include <sddf/timer/protocol.h>
+#include <sddf/util/util.h>
+#include <sddf/util/printf.h>
+#include <sddf/resources/device.h>
+
+__attribute__((__section__(".device_resources"))) device_resources_t device_resources;
+
+#define CLIENT_CH_START 2
+#define MAX_TIMEOUTS 6
+
+#define RK3568_TIMER_CONTROL_TIMER_ENABLE BIT(0)
+#define RK3568_TIMER_CONTROL_MODE_USER BIT(1)
+#define RK3568_TIMER_CONTROL_INTERRUPT_ENABLE BIT(2)
+
+/* 24 MHz frequency. */
+#define RK3568_TIMER_TICKS_PER_SECOND 0x16e3600
+
+typedef struct {
+    uint32_t load_count0;
+    uint32_t load_count1;
+    uint32_t current_value0;
+    uint32_t current_value1;
+    uint32_t control_reg;
+    uint32_t reserved0;
+    uint32_t int_status;
+    uint32_t reserved1;
+} rk3568_timer_regs_t;
+
+static volatile rk3568_timer_regs_t * timestamp_timer;
+static volatile rk3568_timer_regs_t * timeout_timer;
+sddf_channel timestamp_irq;
+sddf_channel timeout_irq;
+
+static uint64_t timeouts[MAX_TIMEOUTS];
+
+static void print_regs(volatile rk3568_timer_regs_t * timer) {
+    sddf_dprintf("TIMER DRIVER|LOG: regs load_count0 : 0x%x\n", timer->load_count0);
+    sddf_dprintf("TIMER DRIVER|LOG: regs load_count1 : 0x%x\n", timer->load_count1);
+    sddf_dprintf("TIMER DRIVER|LOG: regs current_value0: 0x%x\n", timer->current_value0);
+    sddf_dprintf("TIMER DRIVER|LOG: regs current_value1: 0x%x\n", timer->current_value1);
+    sddf_dprintf("TIMER DRIVER|LOG: regs control_reg: 0x%x\n", timer->control_reg);
+    sddf_dprintf("TIMER DRIVER|LOG: regs int_status: 0x%x\n", timer->int_status);
+}
+
+static inline uint64_t get_ticks_in_ns(void)
+{
+    /* the timer value counts down from the load value */
+    uint64_t load_values = timestamp_timer->current_value0;
+    load_values |= (uint64_t)timestamp_timer->current_value1 << 32;
+    uint64_t ticks = UINT64_MAX - load_values;
+
+    /* convert from ticks to nanoseconds */
+    uint64_t value_whole_seconds = ticks / RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ticks = ticks % RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ns = (value_subsecond_ticks * NS_IN_S) / RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t value_ns = value_whole_seconds * NS_IN_S + value_subsecond_ns;
+
+    //sddf_dprintf("TIMER DRIVER|LOG: get_ticks_in_ns load_values: %lu value_ns: %lu\n", load_values, value_ns);
+    return value_ns;
+}
+
+void set_timeout(uint64_t ns)
+{
+    //sddf_dprintf("TIMER DRIVER|LOG: TIMEOUT SET_TIMEOUT\n");
+    //print_regs(timeout_timer);
+    /* load the timeout timer with ticks to count down from */
+    uint64_t ticks_whole_seconds = (ns / NS_IN_S) * RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t ticks_remainder = (ns % NS_IN_S) * RK3568_TIMER_TICKS_PER_SECOND / NS_IN_S ;
+    uint64_t num_ticks = ticks_whole_seconds + ticks_remainder;
```

Comment:
See https://github.com/au-ts/sddf/pull/499#discussion_r2476413522 - this type of calculation is inaccurate and takes more operations than needed.

Reply:
Fixed